### ISSUE TITLE
setting c.Spawner.disable_user_config = True prevents kernelspecs in $HOME from displaying

### ISSUE THREAD

**Issue by kellyrowland** [jupyterhub/jupyterhub#4940]
Setting `c.Spawner.disable_user_config = True` in `jupyterhub_config.py` appears to be preventing kernels installed in `$HOME` (e.g., via `python -m ipykernel install --user`) from displaying.

I would not expect a "config" option to impact kernels, based on

>Jupyter separates data files (nbextensions, kernelspecs) from runtime files (logs, pid files, connection files) from configuration (config files, custom.js).

stated at https://docs.jupyter.org/en/stable/use/jupyter-directories.html so maybe this is an unintended side effect of disabling the user config or maybe kernelspecs should be considered configuration.

**Comment by minrk** [comment]
Yeah, these things are not well separated. disable-user-config removes home directories from the prefixes Jupyter searches for everything, not just config. I'll need to do some digging on exactly where these things differentiate to allow it. I think the issue is that `jupyter_path` which resolves to 'data' directories is also where _extensions_ reside (browser-side ones, at least), so it may still be correct to exclude them. I'll have to think about it some more. While I think kernel specs in particular from $HOME are considered safe, I'm less sure about `data_dir` generally.

I expected `KernelSpecManager.kernel_dirs` to have an override to be able to explicitly re-add a directory, but it's actually not configurable. There isn't a way to re-add a directory without getting your own Python code in there as a sysadmin, e.g. registering a `ServerApp.kernel_spec_manager_class` that customizes `.kernel_dirs`, or an ExtensionApp that accesses the `serverapp.kernel_spec_manager.kernel_dirs` list and appends to it. I will look into making that configurable, at least.

**Comment by mahendrapaipuri** [comment]
> I'm less sure about data_dir generally.

@minrk If the `data_dir` exists outside user's home directory, using `JUPYTER_PATH` env variable user will be still able to load browser side JS files. This can happen on HPC systems where users will have dedicated `HOME`, `WORK`, `SCRATCH` spaces. I think probably we will need to document a bit more on that sanitization of env variable `JUPYTER_PATH` if operators do not want to allow user config. What do you think?

Regarding kernels, `KernelSpecManager` still looks for kernels in `~/.ipython/kernels` directory which can be used to load user kernels when `c.Spawner.disable_user_config = True` is used. It is more of a pragmatic workaround rather than an elegant solution though!
